### PR TITLE
RESTAKING + VAULT: Adding Metadata Admin Field

### PR DESCRIPTION
**Pull request by coachchucksol**
In the future we will have a metadata account for the vaults, operators and ncns. I think it makes sense to keep the admin to update this metadata within it's respective account and this keeps our pattern of separation of authority.

Side effect: This metadata_admin consumed the `reserve` buffer in the middle of the accounts. Which is good to clean it up a bit

### CODE REVIEW

**Comment by coachchucksol** [comment]
I added in two additional ncn admins: `weight_table_admin` and `ncn_program_admin`

I am anticipating the needs of the weight_table and the bespoke ncn on-chain's program's need to be verifiably tied to the ncn struct.

**Review by ebatsell — APPROVED** [review]

File: vault_core/src/vault.rs [hunk 1]
```diff
@@ -520,6 +524,16 @@ impl Vault {
             self.fee_admin = *new_admin;
             msg!("Fee admin set to {:?}", new_admin);
         }
+
+        if self.metadata_admin.eq(old_admin) {
+            self.metadata_admin = *new_admin;
+            msg!("Metadata admin set to {:?}", new_admin);
+        }
+
+        if self.metadata_admin.eq(old_admin) {
```
Comment (buffalu) [inline vault_core/src/vault.rs:533]: extra copy paste

File: restaking_sdk/src/instruction.rs [hunk 2]
```diff
@@ -195,10 +195,13 @@ pub enum RestakingInstruction {
...
-    Operator,
-    Vault,
-    Slasher,
-    Delegate,
+    OperatorAdmin,
+    VaultAdmin,
+    SlasherAdmin,
+    DelegateAdmin,
+    MetadataAdmin,
+    WeightTableAdmin,
```
Comment (buffalu) [inline restaking_sdk/src/instruction.rs:203]: remove this + ncn one i think

**Review by buffalu — COMMENTED** [review]
are you imagining its another account or same account for metadata?

File: restaking_core/src/ncn.rs [hunk 3]
```diff
@@ -33,8 +33,14 @@ pub struct Ncn {
     /// The delegate admin of the NCN
     pub delegate_admin: Pubkey,
 
-    // Reserved space
-    reserved_1: [u8; 32],
+    /// ( For future use ) Authority to update the ncn's metadata
+    pub metadata_admin: Pubkey,
+
+    /// The weight table admin of the NCN
+    pub weight_table_admin: Pubkey,
```
Comment (buffalu) [inline restaking_core/src/ncn.rs:40]: let's remove this - will be diff program. can make choice on which thing is singer later and/or add

File: restaking_core/src/ncn.rs [hunk 4]
```diff
@@ -33,8 +33,14 @@ pub struct Ncn {
...
+
+    /// Admin in charge of of any on-chain programs related to the NCN
+    pub ncn_program_admin: Pubkey,
```
Comment (buffalu) [inline restaking_core/src/ncn.rs:43]: and can remove this for now

**Review by buffalu — APPROVED** [review]

File: restaking_sdk/src/instruction.rs (hunk 2 above)
Comment (coachchucksol) [reply restaking_sdk/src/instruction.rs:203]: Keeping for now

**Comment by coachchucksol** [comment]
To capture the design decision:

> We are going to have a separate metadata program ( much like the validators ), and to make the equivalent of an ATA. For this, we need the NCN to be a signer. And it makes sense to me for that to be a separate admin, such that something low stakes like changing the metadata does not escalate to the main admin for scenarios when each admin can be a multi-sig. So we add the metadata_admin.
> I believe this also extends to the weight table and on-chain ncn programs (ncn_program_admin) . I think it makes sense to have a signer within the NCN to manage these/prove ownership over. This allows us to have accounts address derived from the base NCN, it gives us a single point of control, and generally gives developers a super easy way to prove ownership over the NCN.
> I think every NCN will want a weight table, such that they can determine consensus. And almost all NCNs will have some on-chain component. The ability to prove ownership, or give control to other parties from the ncn account would be super helpful and can make life easier for NCN developers to have that separation of power. Otherwise, they'd have to use the top level admin.
> What these two fields unlock vs the cost of 64 bytes seems worth it to me. I know you don't want to be changing things up, but I think these additions would be helpful.
> This all being said, I've added these in there to support my creation of the Rewards NCN. I can certainly work around it if you want these removed, and am okay with any decision. I figured I'd just share some of my reasoning around it.

File: restaking_core/src/ncn.rs (hunk 4 above)
Comment (coachchucksol) [reply restaking_core/src/ncn.rs:43]: keeping for now

File: restaking_core/src/ncn.rs (hunk 3 above)
Comment (coachchucksol) [reply restaking_core/src/ncn.rs:40]: keeping for now

**Pull request merged**